### DISCUSSION TITLE
Generic method at call as server

### DISCUSSION THREAD

**Discussion by BenediktEnger** [FreeOpcUa/python-opcua#1460] [Q&A] [answered]
Hi,
I try to implement an generic approach to collect called methods. 
When adding a method you habe to enter the id, name , function and arguments.
So as a function I want to use the same for multiple methods. The problem is, the called function does not get any information about who has called. I only get information about the parent but in my case, the parent has multiple methods.

Does anybody knows how do deal with it?

Thanks in advance!

**Comment by AndreasHeine** [comment, marked as answer]
there is no session-context in the method-service 
https://github.com/FreeOpcUa/python-opcua/blob/67f15551884d7f11659d52483e7b932999ca3a75/opcua/server/address_space.py#L456

for what do you need the session info?

**Reply by BenediktEnger** [reply]
I want to create an OPC UA Service which creates the model. The user of the service can create a callback from that method to implement a custom response. To reduce the amount of programming when creating the model, I want to use one generic method, checking for a matching callback. For this, the called method needs to know the name of the method

**Reply by AndreasHeine** [reply]
that sounds weird to me... 
the only way would be to get the whole "CallMethodRequest"-Data inside of the method itself... 
https://github.com/FreeOpcUa/python-opcua/blob/67f15551884d7f11659d52483e7b932999ca3a75/opcua/ua/uaprotocol_auto.py#L8288

**Reply by BenediktEnger** [reply]
See my answer below. I want to create a mock, which by default, creates a random response. If the user of the mock wants to add some workflows or stuff he can get a callback from that method to override the return value. This way it is easy to create and expand the model.
Sure, I could create a single function for each method but it would be much quicker and reduce code if I could go the generic way.

Can you give me a hint, how to get this data into the method? I guess I have to override repo code, right?

**Reply by AndreasHeine** [reply]
> Can you give me a hint, how to get this data into the method? I guess I have to override repo code, right?

for sure!

you need to modify internal_session, callservice, ... (passing all that "CallMethodRequest"-data along! till the inputs of the ua method itself -> like method(request, parent, *args) )

**Comment by swamper123** [comment]
Did I understand that correctly, that you want to make one OPC UA method which shall execute a function depending on the input arguments: ID, name, function and *args ?

**Reply by BenediktEnger** [reply]
yes, I created a super class for an object which has the called method in it. sub classes of it can implement a callback on that method to generic a custom response. This is used for as a mock of a real system
